src/iop/channelmixerrgb.c

```diff
@@ -735,6 +743,9 @@ static inline void loop_switch(const float *const restrict in, float *const rest
         // Convert from RGB to XYZ
         dot_product(temp_one, RGB_to_XYZ, temp_two);
 
+        #ifdef _OPENMP
+        #pragma omp simd aligned(temp_one, temp_two:16)
+        #endif
         for(size_t c = 0; c < 3; ++c) temp_one[c] = temp_two[c];
```

Comment:
GCC doesn't actually vectorize if you only iterate over three elements (it needs to be four).  So there's likely to be a bit more speedup possible.
`for_each_channel` would be appropriate here and the other three-item loops where adding the fourth doesn't harm anything (the loops with reductions probably need to stay at three elements).

Reply:
I know, but the goal is to let the compiler heuristics decide if it's worth it to vectorize or not. I have tested the runtimes with 3 and 4-elements loops, and it runs faster this way on Xeon. But it depends on the loops so I tested each of them separately…

Reply:
This is a case where one needs to look at the generated assembly to find out why vectorized is slower than scalar (because IME GCC will never vectorize a three-iteration loop).  My guess: your Xeon supports AVX-512, GCC generates such instructions when you use a four elements for some of the loops in this code, and you're getting hit by the well-known 30% clock-throttling for using AVX-512 -- which continues for several milliseconds afterwards.  So your actual loops may be running faster, but the side-effect slows down the overall code.

What is the performance like if you iterate four elements but add `safelen(4)` or `safelen(8)` to the OpenMP directive?  That would limit the code generation to 128-bit resp. 256-bit instructions.  If that generates overall faster code, we can adjust the definition of `for_each_loop` to take that into account, possible even on an architecture-by-architecture basis.